daemon/src/GHCSpecter/Control.hs

```diff
@@ -0,0 +1,251 @@
+          pure (newUI', newSS)
+        SubModuleEv sev ->
+          case sev of
+            SubModuleGraphEv ev -> do
+              let mgui = oldUI ^. uiSubModuleGraph . _2
+              (mgui', mxy) <- handleModuleGraphEv ev mgui
+              let newUI = (uiSubModuleGraph . _2 .~ mgui') oldUI
+                  newUI' = handleMouseMove newUI mxy
+                  newSS = (serverShouldUpdate .~ False) oldSS
+              pure (newUI', newSS)
+            SubModuleLevelEv d' -> do
+              let newUI = (uiSubModuleGraph . _1 .~ d') oldUI
+                  newSS = (serverShouldUpdate .~ False) oldSS
+              pure (newUI, newSS)
+        SessionEv SaveSessionEv -> do
+          saveSession
+          let newSS = (serverShouldUpdate .~ False) oldSS
+          pure (oldUI, newSS)
+        SessionEv ResumeSessionEv -> do
+          let sinfo = oldSS ^. serverSessionInfo
+              sinfo' = sinfo {sessionIsPaused = False}
+              newSS = (serverSessionInfo .~ sinfo') . (serverShouldUpdate .~ True) $ oldSS
+          pure (oldUI, newSS)
+        SessionEv PauseSessionEv -> do
+          let sinfo = oldSS ^. serverSessionInfo
+              sinfo' = sinfo {sessionIsPaused = True}
+              newSS = (serverSessionInfo .~ sinfo') . (serverShouldUpdate .~ True) $ oldSS
+          pure (oldUI, newSS)
+        TimingEv (UpdateSticky b) -> do
+          let newUI = (uiTiming . timingUISticky .~ b) oldUI
+              newSS = (serverShouldUpdate .~ False) oldSS
+          pure (newUI, newSS)
+        TimingEv (UpdatePartition b) -> do
+          let newUI = (uiTiming . timingUIPartition .~ b) oldUI
+              newSS = (serverShouldUpdate .~ False) oldSS
+          pure (newUI, newSS)
+        TimingEv (UpdateParallel b) -> do
+          let newUI = (uiTiming . timingUIHowParallel .~ b) oldUI
+              newSS = (serverShouldUpdate .~ False) oldSS
+          pure (newUI, newSS)
+        MessageChanUpdated -> do
+          let newSS = (serverShouldUpdate .~ True) oldSS
+          pure (oldUI, newSS)
+
+    handleModuleGraphEv ::
+      ModuleGraphEvent ->
+      ModuleGraphUI ->
+      Control (ModuleGraphUI, Maybe (UTCTime, (Double, Double)))
+    handleModuleGraphEv (HoverOnModuleEv mhovered) mgui =
+      pure ((modGraphUIHover .~ mhovered) mgui, Nothing)
+    handleModuleGraphEv (ClickOnModuleEv mclicked) mgui =
+      pure ((modGraphUIClick .~ mclicked) mgui, Nothing)
+    handleModuleGraphEv (DummyEv mxy) mgui = do
+      t <- getCurrentTime
+      printMsg (T.pack (show (stepStartTime, t, mxy)))
+      pure (mgui, (t,) <$> mxy)
+
+    handleMouseMove ::
+      UIState ->
+      Maybe (UTCTime, (Double, Double)) ->
+      UIState
+    handleMouseMove ui_ mtxy =
+      case mtxy of
+        Nothing -> ui_
+        Just (t, xy) ->
+          if ui_ ^. uiShouldUpdate
+            then (uiLastUpdated .~ t) . (uiMousePosition .~ xy) $ ui_
+            else uiMousePosition .~ xy $ ui_
+
+control :: Control ()
```

Comment:
This is an actual user code while the other parts of this module are library code. We had better separate this module to two.

Reply:
Now separated out Control.Types and Control.Runner and this module with only user's program logic function.